### ISSUE TITLE
Trust Policy check counts Functions in length (E2508 Role trust policy JSON text cannot be longer than 2048 characters)

### ISSUE THREAD

**Issue by ikben** [aws-cloudformation/cfn-python-lint#738]
*cfn-lint version: (`cfn-lint --version`)*
cfn-lint 0.16.0

*Description of issue.*
When using functions in a Trust Policy, they get counted as characters and trigger the length check. This may cause false positives eg, if there is an "If" function in there, or false negatives, eg if the there is a Ref('short') that returns a long value.
 
Example template:
```
Parameters:
  # Testing with this set to No will deploy
  LimitAdminUse:
    AllowedValues:
      - 'Yes'
      - 'No'
    Default: 'No'
    Type: String
  UserName:
    Type: String
Conditions:
  IsLimitAdminUse: !Equals
    - !Ref 'LimitAdminUse'
    - 'Yes'
Resources:
  AdminRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: sts:AssumeRole
            Condition:
              Bool:
                aws:MultiFactorAuthPresent: 'true'
            Effect: Allow
            Principal:
              AWS: !If
                - IsLimitAdminUse
                # We repeat a lot, to get more characters, this could be all different usernames in a real template
                - - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                  - !Sub
                    - arn:aws:iam::${Account}:user/${User}
                    - Account: !Ref AWS::AccountId
                      User: !Ref UserName
                - !Sub
                  - arn:aws:iam::${Account}:root
                  - Account: !Ref AWS::AccountId
            Sid: ''
        Version: '2012-10-17'
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/AdministratorAccess
      Path: /
```

* Feature request:
Getting this check to work reliably is hard, there are some mays to make it better.

1. Ignore the check if there are functions in the trust policy
2. (a bit better) Count every function as 0 characters (this should return the minimum number of characters)
2. (even better). Count every function that returns a value as 0, and handle If's as the maximum of its branches.

**Comment by kddejong** [comment]
agreed.  I've been doing a lot of work on how to process conditions.  I think I have enough of the core engine built to do option 3.  I still need to do a few things so give me a little time

**Comment by kddejong** [comment]
I got the code to fix this working.  I'm going to have to test it a little bit more and make sure I covered most of the scenarios.   This will work as you have describe in option 3.  

1. We will remove the Fn::Ifs by playing out the condition logic.
1. Then we will remove all other functions (besides Fn::Sub) replacing them with 0.  For Fn::Sub we will use the string but replace any variables with an empty string.
1. Then we will check the size of the resulting JSON

This way we can expand the logic around Fn::Sub to get even closer to reality.  Replace AccountId with a 12 digit number, etc.

**Comment by ikben** [comment]
That's awesome!

May I suggest treating Fn::Join similar to Fn::Sub (keeping the length of anything that is a real string, and maybe in the future count Ref's to an AccountId as 12).

The `Fn::Base64` is also pretty straight-forward to count, as this is always at least the length of the string given to it. More accurately it's the length of the string * 8\6 (8 bits per string characters, and 6 bits per base64 character)

The other functions are harder to do :)
